analyzer/emerald/incoming.go

```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
+          "name": "_owner",
+          "type": "address"
+        },
+        {
+          "indexed": true,
+          "name": "_spender",
+          "type": "address"
+        },
+        {
+          "indexed": false,
+          "name": "_value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Approval",
+      "type": "event"
+    }
+  ]
+`))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// nolint: unparam
+func evmCallWithABI(
+	ctx context.Context,
+	client *oasis.RuntimeClient,
+	round uint64,
+	contractEthAddr []byte,
+	contractABI *abi.ABI,
+	name string,
+	outI interface{},
+	in ...interface{},
+) error {
+	// https://github.com/oasisprotocol/emerald-web3-gateway/blob/v3.0.0/rpc/eth/api.go#L408
+	gasPrice := []byte{1}
+	gasLimit := uint64(30_000_000)
+	caller := ethCommon.Address{1}.Bytes()
+	value := []byte{0}
+
+	inPacked, err := contractABI.Pack(name, in...)
+	if err != nil {
+		panic(err)
+	}
+	outPacked, err := client.EvmSimulateCall(ctx, round, gasPrice, gasLimit, caller, contractEthAddr, value, inPacked)
+	if err != nil {
+		return fmt.Errorf("runtime client evm simulate call: %w", err)
+	}
+	if err = contractABI.UnpackIntoInterface(outI, name, outPacked); err != nil {
+		return fmt.Errorf("unpack out: %w", err)
+	}
+	return nil
+}
+
+func downloadTokenERC20(ctx context.Context, logger *log.Logger, client *oasis.RuntimeClient, round uint64, tokenEthAddr []byte) (*TokenData, error) {
+	var tokenData TokenData
+	logError := func(method string, err error) {
+		logger.Info("ERC20 call failed",
+			"round", round,
+			"token_eth_addr_hex", hex.EncodeToString(tokenEthAddr),
+			"method", method,
+			"err", err,
+		)
+	}
+	if err := evmCallWithABI(ctx, client, round, tokenEthAddr, &ABIERC20, "name", &tokenData.Name); err != nil {
+		// todo: propagate nondeterministic errors
+		logError("name", err)
+	}
+	if err := evmCallWithABI(ctx, client, round, tokenEthAddr, &ABIERC20, "symbol", &tokenData.Symbol); err != nil {
+		logError("symbol", err)
+	}
+	if err := evmCallWithABI(ctx, client, round, tokenEthAddr, &ABIERC20, "decimals", &tokenData.Decimals); err != nil {
+		logError("decimals", err)
+	}
+	if err := evmCallWithABI(ctx, client, round, tokenEthAddr, &ABIERC20, "totalSupply", &tokenData.TotalSupply); err != nil {
+		logError("totalSupply", err)
+	}
+	return &tokenData, nil
+}
+
+func downloadRoundTokens(ctx context.Context, logger *log.Logger, client *oasis.RuntimeClient, round uint64, tokens map[string]*DirtyToken) (*BlockTokenData, error) {
+	var blockTokenData BlockTokenData
+	blockTokenData.TokenData = map[string]*TokenData{}
+	// todo: slow and serialized, lots of round trips
+	for tokenAddr, dirtyToken := range tokens {
+		switch dirtyToken.Type {
+		case TokenTypeERC20:
+			tokenData, err := downloadTokenERC20(ctx, logger, client, round, dirtyToken.EthAddr)
+			if err != nil {
+				return nil, fmt.Errorf("download token ERC20 %s %s (eth): %w", tokenAddr, hex.EncodeToString(dirtyToken.EthAddr), err)
```

Comment:
"we found that it doesn't really implement this token standard" (A) is one outcome, and we'd go on to try other candidates or mark the token as unrecognized. however, and this may be a minor point, I'm not decided on using a Go error return to indicate that case.

other outcomes that'll be important:

"this was an ERC20, and we have its data" (B)

"this was an ERC20, but some optional methods weren't supported" (C)

"a network error prevented us from accessing this contract" (D)

D will be a Go error. A might or might not. C is close enough to B that we can likely have a struct with nullable fields.

Reply:
D we'll want to break out of this loop, unless maybe we want to retain results and save partial data to the db. we usually do this at larger granularity though e.g. per block in the scanner.

to handle A as an error, we might be able to play around with error.Is et al. or it might be simpler to have it be a kind of success

Reply:
I think I wasn't very clear in my original comment; I was (also?) advocating for this not being a Go error in case A. Good point that D exists and is a very different failure. Yeah I agree with the `// todo` that nondeterministic errors (meaning type D) should be propagated out of `evmDownloadTokenERC20()`, and other errors should just impact setting the token type. IIRC you said in a different comment thread that you want to do that in a separate PR? Either way.

Reply:
will be carrying on in a separate PR. thanks for the ideas. resolving for merge